### PR TITLE
Drop usage of internal clients, informers and listers

### PR DESCRIPTION
**Pull request by ary1992**
**How to categorize this PR?**

/area open-source
/kind enhancement

**What this PR does / why we need it**:
Kubernetes code-generator does not generate clients, listers and informers for internal packages. With this PR we dropping the usage internal clients, informers and lister. Instead now we will be using versioned clients, internal informers and listers.

**Which issue(s) this PR fixes**:
Part of https://github.com/gardener/gardener/issues/8382

**Special notes for your reviewer**:
~~In draft as few minor adaptations are left out. The PR is open for suggestions though.~~

**Release note**:

```other operator
NONE
```

### CODE REVIEW

File: hack/update-codegen.sh [hunk 1]
```diff
@@ -133,9 +137,10 @@ export -f resources_groups
 operator_groups() {
   echo "Generating API groups for pkg/apis/operator"
 
-  bash "${PROJECT_ROOT}"/hack/generate-groups.sh \
+  bash "${CODE_GEN_DIR}"/generate-internal-groups.sh \
```
Comment (shafeeqes) [inline hack/update-codegen.sh:140]: Can we just call `generate-groups.sh` itself, I see they take care of passing `""` internally.

File: pkg/apis/core/helper/helpers.go [hunk 2]
```diff
@@ -525,7 +525,7 @@ func IsWorkerless(shoot *core.Shoot) bool {
 }
 
 // ShootEnablesSSHAccess returns true if ssh access to worker nodes should be allowed for the given shoot.
-func ShootEnablesSSHAccess(shoot *core.Shoot) bool {
+func ShootEnablesSSHAccess(shoot *gardencorev1beta1.Shoot) bool {
```
Comment (shafeeqes) [inline pkg/apis/core/helper/helpers.go]: Why is this change needed, we already have https://github.com/gardener/gardener/blob/ae123edaf67fea5389ccc16514b6a319d031ae03/pkg/apis/core/v1beta1/helper/helper.go#L1475-L1479

File: plugin/pkg/utils/miscellaneous.go [hunk 3]
```diff
@@ -24,7 +24,8 @@ import (
 	"k8s.io/apiserver/pkg/admission"
 
 	"github.com/gardener/gardener/pkg/apis/core"
-	gardencorelisters "github.com/gardener/gardener/pkg/client/core/listers/core/internalversion"
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	gardencorelisters "github.com/gardener/gardener/pkg/client/core/listers/core/v1beta1"
```
Comment (shafeeqes) [inline plugin/pkg/utils/miscellaneous.go]: Should this not be called `gardencorev1beta1listers`?
Similarly in other places.

File: pkg/apis/core/helper/helpers.go [hunk 4]
```diff
@@ -164,7 +164,7 @@ func TaintsHave(taints []core.SeedTaint, key string) bool {
 }
 
 // TaintsAreTolerated returns true when all the given taints are tolerated by the given tolerations.
-func TaintsAreTolerated(taints []core.SeedTaint, tolerations []core.Toleration) bool {
+func TaintsAreTolerated(taints []gardencorev1beta1.SeedTaint, tolerations []core.Toleration) bool {
```
Comment (shafeeqes) [inline pkg/apis/core/helper/helpers.go]: Please move this to v1beta1 helpers.
And similarly for others.

File: cmd/gardener-apiserver/app/gardener_apiserver.go [hunk 5]
```diff
@@ -114,8 +112,7 @@ type Options struct {
 	Recommended                   *genericoptions.RecommendedOptions
 	ServerRunOptions              *genericoptions.ServerRunOptions
 	ExtraOptions                  *apiserver.ExtraOptions
-	CoreInformerFactory           gardencoreinformers.SharedInformerFactory
-	ExternalCoreInformerFactory   gardencoreexternalinformers.SharedInformerFactory
+	ExternalCoreInformerFactory   gardencoreinformers.SharedInformerFactory
```
Comment (shafeeqes) [inline cmd/gardener-apiserver/app/gardener_apiserver.go]: Can we call this `CoreInformerFactory` now since we only have one kind?

File: cmd/gardener-apiserver/app/gardener_apiserver.go [hunk 6]
```diff
@@ -199,20 +196,13 @@ func (o *Options) config(kubeAPIServerConfig *rest.Config, kubeClient *kubernete
...
-	coreClient, err := gardencoreclientset.NewForConfig(&protobufLoopbackConfig)
-	if err != nil {
-		return nil, err
-	}
-	o.CoreInformerFactory = gardencoreinformers.NewSharedInformerFactory(coreClient, protobufLoopbackConfig.Timeout)
-	apiConfig.CoreInformerFactory = o.CoreInformerFactory
-
 	// versioned core client
-	versionedCoreClient, err := gardencoreversionedclientset.NewForConfig(&protobufLoopbackConfig)
+	versionedCoreClient, err := gardencoreclientset.NewForConfig(&protobufLoopbackConfig)
```
Comment (shafeeqes) [inline cmd/gardener-apiserver/app/gardener_apiserver.go]: Can we call this `coreClient` now?

File: .golangci.yaml [hunk 7]
```diff
@@ -172,15 +172,11 @@ linters-settings:
...
       - pkg: github.com/gardener/gardener/pkg/client/core/clientset/versioned
-        alias: gardencoreversionedclientset
+        alias: gardencoreclientset
       - pkg: github.com/gardener/gardener/pkg/client/core/informers/externalversions
-        alias: gardencoreexternalinformers
-      - pkg: github.com/gardener/gardener/pkg/client/core/informers/internalversion
         alias: gardencoreinformers
-      - pkg: github.com/gardener/gardener/pkg/client/core/listers/core/internalversion
+      - pkg: github.com/gardener/gardener/pkg/client/core/listers/core/v1beta1
         alias: gardencorelisters
```
Comment (shafeeqes) [inline .golangci.yaml]:
```suggestion
        alias: gardencorev1beta1listers
```

**Review by shafeeqes — CHANGES_REQUESTED** [review]
Thanks for the huge effort 👏 
It's well structured.

File: plugin/pkg/utils/miscellaneous.go (hunk 3 above)
Comment (ary1992) [reply plugin/pkg/utils/miscellaneous.go]: As we no longer have the internal listers and informers. I chose to name it it `gardencorelisters`. WDYT?

Comment (shafeeqes) [reply plugin/pkg/utils/miscellaneous.go]: I think for consistency reasons, we should use v1beta1 in the name. What if in the future there is another version?

File: .golangci.yaml (hunk 7 above)
Comment (ary1992) [reply .golangci.yaml]: We have only one listers now right. I think its better to just keep it as `gardencorelisters`

Comment (shafeeqes) [reply .golangci.yaml]: Same argument as in https://github.com/gardener/gardener/pull/8851#discussion_r1436746245

File: hack/update-codegen.sh (hunk 1 above)
Comment (ary1992) [reply hack/update-codegen.sh:140]: On using `generate-group.sh`, I am getting `permission denied` error
```
go/pkg/mod/k8s.io/code-generator@v0.28.3/generate-groups.sh: line 66: go/pkg/mod/k8s.io/code-generator@v0.28.3/generate-internal-groups.sh: Permission denied
```
because `generate-groups.sh` is using `exec` call to call the `generate-internal-groups.sh`
https://github.com/kubernetes/kubernetes/blob/4909dbd0d7fd1ca13dc8c8db3af3cbf4733f3035/staging/src/k8s.io/code-generator/generate-groups.sh#L65-L66

I would suggest lets proceed with the current change as we will migrate to new `kube-codegen` script from the next release of kubernetes i.e `1.30`

File: pkg/apis/core/helper/helpers.go (hunk 2 above)
Comment (ary1992) [reply pkg/apis/core/helper/helpers.go]: It was changed back in the later commit.

File: pkg/apis/core/helper/helpers.go [hunk 8]
```diff
@@ -164,7 +164,7 @@ func TaintsHave(taints []core.SeedTaint, key string) bool {
 }
 
 // TaintsAreTolerated returns true when all the given taints are tolerated by the given tolerations.
-func TaintsAreTolerated(taints []gardencorev1beta1.SeedTaint, tolerations []core.Toleration) bool {
+func TaintsAreTolerated(taints []core.SeedTaint, tolerations []core.Toleration) bool {
```
Comment (shafeeqes) [inline pkg/apis/core/helper/helpers.go]: Can you remove this and other functions in this file which are not used anymore?

Comment (ary1992) [reply pkg/apis/core/helper/helpers.go]: There are other functions as well which are unused in our code base we need to find them all and remove it. We should do the cleanup part in a separate PR. I have removed the functions which are related to this PR.

Comment (shafeeqes) [reply pkg/apis/core/helper/helpers.go]: I meant just this file.

File: cmd/gardener-apiserver/app/gardener_apiserver.go [hunk 9]
```diff
@@ -199,21 +196,14 @@ func (o *Options) config(kubeAPIServerConfig *rest.Config, kubeClient *kubernete
 		protobufLoopbackConfig.ContentType = runtime.ContentTypeProtobuf
 	}
 
-	// core client
+	// versioned core client
```
Comment (shafeeqes) [inline cmd/gardener-apiserver/app/gardener_apiserver.go]:
```suggestion
```
Now that we only have one client

File: pkg/apiserver/admission/initializer/types.go [hunk 10]
```diff
@@ -22,36 +22,22 @@ import (
...
 type WantsExternalCoreInformerFactory interface {
-	SetExternalCoreInformerFactory(gardencoreexternalinformers.SharedInformerFactory)
+	SetExternalCoreInformerFactory(gardencoreinformers.SharedInformerFactory)
 	admission.InitializationValidator
 }
 
-// WantsExternalCoreClientset defines a function which sets external Core Clientset for admission plugins that need it.
-type WantsExternalCoreClientset interface {
-	SetExternalCoreClientset(gardencoreversionedclientset.Interface)
+// WantsExternalCoreClientSet defines a function which sets external Core Clientset for admission plugins that need it.
```
Comment (shafeeqes) [inline pkg/apiserver/admission/initializer/types.go]: Do we still need to call them as External? We only have one now. What does Kubernetes do?

File: plugin/pkg/bastion/validator/admission.go [hunk 11]
```diff
@@ -151,7 +157,7 @@ func (v *Bastion) Admit(ctx context.Context, a admission.Attributes, _ admission
 	}
 
 	// ensure shoot SSH access is not disabled
-	if bastion.DeletionTimestamp == nil && !gardencorehelper.ShootEnablesSSHAccess(shoot) {
+	if bastion.DeletionTimestamp == nil && !gardencorehelper.ShootEnablesSSHAccess(&coreShoot) {
```
Comment (shafeeqes) [inline plugin/pkg/bastion/validator/admission.go]: If this is the only place `coreShoot` is used, we don't need to convert the object. You can call `v1beta1helper.ShootEnablesSSHAccess`

File: plugin/pkg/managedseed/validator/admission.go [hunk 12]
```diff
@@ -201,17 +202,22 @@ func (v *ManagedSeed) Admit(ctx context.Context, a admission.Attributes, _ admis
...
+	if err := gardencorev1beta1.Convert_v1beta1_Shoot_To_core_Shoot(shoot, coreShoot, nil); err != nil {
+		return err
+	}
+
 	// Ensure shoot can be registered as seed
 	if shoot.Spec.DNS == nil || shoot.Spec.DNS.Domain == nil || *shoot.Spec.DNS.Domain == "" {
 		return apierrors.NewInvalid(gk, managedSeed.Name, append(allErrs, field.Invalid(shootNamePath, managedSeed.Spec.Shoot.Name, fmt.Sprintf("shoot %s does not specify a domain", kubernetesutils.ObjectName(shoot)))))
 	}
-	if gardencorehelper.NginxIngressEnabled(shoot.Spec.Addons) {
+	if gardencorehelper.NginxIngressEnabled(coreShoot.Spec.Addons) {
```
Comment (shafeeqes) [inline plugin/pkg/managedseed/validator/admission.go]: We have `v1betahelper.NginxIngressEnabled`. Same for other occurences.

File: plugin/pkg/managedseed/validator/admission.go [hunk 13]
```diff
@@ -226,7 +232,7 @@ func (v *ManagedSeed) Admit(ctx context.Context, a admission.Attributes, _ admis
 
 	if managedSeed.Spec.Gardenlet != nil {
 		// Admit gardenlet against shoot
-		errs, err := v.admitGardenlet(managedSeed.Spec.Gardenlet, shoot, field.NewPath("spec", "gardenlet"))
+		errs, err := v.admitGardenlet(managedSeed.Spec.Gardenlet, coreShoot, field.NewPath("spec", "gardenlet"))
```
Comment (shafeeqes) [inline plugin/pkg/managedseed/validator/admission.go]: The function signature can be changed to take a v1beta1 Shoot?

File: plugin/pkg/shoot/quotavalidator/admission.go [hunk 14]
```diff
@@ -332,7 +333,12 @@ func (q *QuotaValidator) findShootsReferQuota(quota core.Quota, shoot core.Shoot
...
 }
 
-func (q *QuotaValidator) determineRequiredResources(allocatedResources corev1.ResourceList, shoot core.Shoot) (corev1.ResourceList, error) {
+func (q *QuotaValidator) determineRequiredResources(allocatedResources corev1.ResourceList, coreShoot core.Shoot) (corev1.ResourceList, error) {
+	shoot := gardencorev1beta1.Shoot{}
+	if err := gardencorev1beta1.Convert_core_Shoot_To_v1beta1_Shoot(&coreShoot, &shoot, nil); err != nil {
+		return nil, err
+	}
+
 	shootResources, err := q.getShootResources(shoot)
```
Comment (shafeeqes) [inline plugin/pkg/shoot/quotavalidator/admission.go]: The core Shoot is converted to version Shoot and then back inside `q.getShootResources`? Can we not pass the core Shoot directly?

File: plugin/pkg/shoot/quotavalidator/admission.go [hunk 15]
```diff
@@ -345,18 +351,28 @@ func (q *QuotaValidator) determineRequiredResources(allocatedResources corev1.Re
...
-func (q *QuotaValidator) getShootResources(shoot core.Shoot) (corev1.ResourceList, error) {
+func (q *QuotaValidator) getShootResources(shoot gardencorev1beta1.Shoot) (corev1.ResourceList, error) {
 	cloudProfile, err := q.cloudProfileLister.Get(shoot.Spec.CloudProfileName)
 	if err != nil {
 		return nil, apierrors.NewInternalError(fmt.Errorf("could not find referenced cloud profile: %+v", err.Error()))
 	}
 
+	coreCloudProfile := core.CloudProfile{}
+	if err := gardencorev1beta1.Convert_v1beta1_CloudProfile_To_core_CloudProfile(cloudProfile, &coreCloudProfile, nil); err != nil {
+		return nil, apierrors.NewInternalError(err)
```
Comment (shafeeqes) [inline plugin/pkg/shoot/quotavalidator/admission.go]:
```suggestion
		return nil, apierrors.NewInternalError(fmt.Errorf("could not convert v1beta1 cloudProfile: %+v", err.Error()))
```
Similarly at other places, if any

File: plugin/pkg/shoot/validator/admission.go [hunk 16]
```diff
@@ -837,7 +850,13 @@ func (c *validationContext) validateProvider(a admission.Attributes) field.Error
...
 
 	if a.GetOperation() == admission.Create && c.secretBinding != nil {
-		if !helper.SecretBindingHasType(c.secretBinding, c.shoot.Spec.Provider.Type) {
+		coreSecretBinding := core.SecretBinding{}
+		if err := gardencorev1beta1.Convert_v1beta1_SecretBinding_To_core_SecretBinding(c.secretBinding, &coreSecretBinding, nil); err != nil {
+			allErrs = append(allErrs, field.InternalError(path.Child("type"), err))
+			return allErrs
+		}
+
+		if !helper.SecretBindingHasType(&coreSecretBinding, c.shoot.Spec.Provider.Type) {
```
Comment (shafeeqes) [inline plugin/pkg/shoot/validator/admission.go]: We have `v1beta1helpers.SecretBindingHasType`. No need to convert

File: plugin/pkg/shoot/validator/admission.go [hunk 17]
```diff
@@ -1296,7 +1321,7 @@ func isUnavailableInAtleastOneZone(regions []core.Region, region string, zones [
 	return false
 }
 
-func validateKubeletConfig(fldPath *field.Path, machineTypes []core.MachineType, workerMachineType string, kubeletConfig *core.KubeletConfig) field.ErrorList {
+func validateKubeletConfig(fldPath *field.Path, machineTypes []gardencorev1beta1.MachineType, workerMachineType string, kubeletConfig *core.KubeletConfig) field.ErrorList {
```
Comment (shafeeqes) [inline plugin/pkg/shoot/validator/admission.go:1312]: Why to change all these functions? You have the core cloudprofile already, Can we not use it everywhere?

**Review by shafeeqes — CHANGES_REQUESTED** [review]

File: pkg/apiserver/admission/initializer/types.go (hunk 10 above)
Comment (ary1992) [reply pkg/apiserver/admission/initializer/types.go]: On kubernetes side I see them using `ExternalCoreCLientSet` 

https://github.com/kubernetes/kubernetes/blob/feb18b3f5f9d443c27dd8cccb6358f271f887744/staging/src/k8s.io/apiserver/pkg/admission/initializer/interfaces.go#L32-L35

Comment (shafeeqes) [reply pkg/apiserver/admission/initializer/types.go]: Okay, then let's keep it as it is.

File: pkg/scheduler/controller/shoot/reconciler.go [hunk 18]
```diff
@@ -347,8 +349,10 @@ func getSeedWithLeastShootsDeployed(seedList []gardencorev1beta1.Seed, shootList
 	var (
 		bestCandidate gardencorev1beta1.Seed
 		min           *int
-		seedUsage     = v1beta1helper.CalculateSeedUsage(shootList)
+		seedUsage     map[string]int
+		sl            = convertList(shootList)
```
Comment (shafeeqes) [inline pkg/scheduler/controller/shoot/reconciler.go]: Nit: Call the incoming variable sl and this variable as shootList

File: plugin/pkg/shoot/validator/admission.go [hunk 19]
```diff
@@ -591,7 +598,13 @@ func (c *validationContext) ensureMachineImages() field.ErrorList {
 	if c.shoot.DeletionTimestamp == nil {
 		for idx, worker := range c.shoot.Spec.Provider.Workers {
 			fldPath := field.NewPath("spec", "provider", "workers").Index(idx)
-			image, err := ensureMachineImage(c.oldShoot.Spec.Provider.Workers, worker, c.cloudProfile.Spec.MachineImages, fldPath)
+			coreCloudProfile := core.CloudProfile{}
+			if err := gardencorev1beta1.Convert_v1beta1_CloudProfile_To_core_CloudProfile(c.cloudProfile, &coreCloudProfile, nil); err != nil {
+				allErrs = append(allErrs, field.InternalError(fldPath, err))
+				return allErrs
+			}
```
Comment (shafeeqes) [inline plugin/pkg/shoot/validator/admission.go]: In the quota validator, you convert the core shoot to versioned shoot, why not be consistent and do the same here instead of converting the cloudprofile?

File: plugin/pkg/shoot/validator/admission.go [hunk 20]
```diff
@@ -755,7 +768,13 @@ func (c *validationContext) validateKubernetes(a admission.Attributes) field.Err
 		return nil
 	}
 
-	defaultVersion, errList := defaultKubernetesVersion(c.cloudProfile.Spec.Kubernetes.Versions, c.shoot.Spec.Kubernetes.Version, path.Child("version"))
+	coreCloudProfile := core.CloudProfile{}
+	if err := gardencorev1beta1.Convert_v1beta1_CloudProfile_To_core_CloudProfile(c.cloudProfile, &coreCloudProfile, nil); err != nil {
+		allErrs = append(allErrs, field.InternalError(path, err))
+		return allErrs
+	}
```
Comment (shafeeqes) [inline plugin/pkg/shoot/validator/admission.go]: same here

File: plugin/pkg/shoot/validator/admission.go [hunk 21]
```diff
@@ -856,6 +875,12 @@ func (c *validationContext) validateProvider(a admission.Attributes) field.Error
 		return allErrs
 	}
 
+	coreCloudProfile := core.CloudProfile{}
+	if err := gardencorev1beta1.Convert_v1beta1_CloudProfile_To_core_CloudProfile(c.cloudProfile, &coreCloudProfile, nil); err != nil {
+		allErrs = append(allErrs, field.InternalError(path, err))
+		return allErrs
+	}
```
Comment (shafeeqes) [inline plugin/pkg/shoot/validator/admission.go]: same here

**Review by shafeeqes — CHANGES_REQUESTED** [review]

**Review by timuthy — COMMENTED** [review]
Generally the changes in the PR are fine, thank you 👍 
As discussed offline, I'd be helpful to have a **common approach** to handle the mishmash of external and internal object versions, either:
- Transform all retrieved objects to the `internal` version first before any checks are performed (- many conversions required, + mimics current approach and means probably less effort when we add support for another API version, e.g. `core/v1`).
- Always work with external version when calling utilities. So instead of calling `pkg/apis/core/helper` functions, convert object from `admission.Attributes` and call helpers from `pkg/apis/core/v1beta1`. (+ less conversion, + less helper duplication in `pkg/apis/core/helper`, - more effort to adjust admission when another version is introduced)

**Comment by ary1992** [comment]
> Transform all retrieved objects to the internal version first before any checks are performed (- many conversions required, + mimics current approach and means probably less effort when we add support for another API version, e.g. core/v1).

I have followed the first approach.

File: pkg/scheduler/controller/shoot/reconciler.go [hunk 22]
```diff
@@ -345,14 +345,14 @@ func filterCandidates(shoot *gardencorev1beta1.Shoot, versionedShootList []garde
 }
 
 // getSeedWithLeastShootsDeployed finds the best candidate (i.e. the one managing the smallest number of shoots right now).
-func getSeedWithLeastShootsDeployed(seedList []gardencorev1beta1.Seed, shootList []gardencorev1beta1.Shoot) (*gardencorev1beta1.Seed, error) {
+func getSeedWithLeastShootsDeployed(seedList []gardencorev1beta1.Seed, versionedShootList []gardencorev1beta1.Shoot) (*gardencorev1beta1.Seed, error) {
```
Comment (timuthy) [inline pkg/scheduler/controller/shoot/reconciler.go]: Nit: Revert this change, it is evident that these are versioned objects since it's in a controller.

File: plugin/pkg/shoot/validator/admission.go [hunk 23]
```diff
@@ -497,13 +521,13 @@ func (c *validationContext) validateScheduling(ctx context.Context, a admission.
...
 
-func getNumberOfShootsOnSeed(shootLister gardencorelisters.ShootLister, seedName string) (int64, error) {
+func getNumberOfShootsOnSeed(shootLister gardencorev1beta1listers.ShootLister, seedName string) (int64, error) {
 	allShoots, err := shootLister.Shoots(metav1.NamespaceAll).List(labels.Everything())
 	if err != nil {
 		return 0, fmt.Errorf("could not list all shoots: %w", err)
 	}
 
-	seedUsage := helper.CalculateSeedUsage(allShoots)
+	seedUsage := v1beta1helper.CalculateSeedUsage(allShoots)
```
Comment (timuthy) [inline plugin/pkg/shoot/validator/admission.go:518]: Shouldn't this be the internal version?

File: plugin/pkg/utils/miscellaneous.go [hunk 24]
```diff
@@ -33,7 +34,7 @@ func SkipVerification(operation admission.Operation, metadata metav1.ObjectMeta)
 }
 
 // IsSeedUsedByShoot checks whether there is a shoot cluster referencing the provided seed name
-func IsSeedUsedByShoot(seedName string, shoots []*core.Shoot) bool {
+func IsSeedUsedByShoot(seedName string, shoots []*gardencorev1beta1.Shoot) bool {
```
Comment (timuthy) [inline plugin/pkg/utils/miscellaneous.go:37]: I didn't expect this change after following the concept of converting to the internal version first. Is this a left over?
Similar concerns for other changes in this file, as well.

File: pkg/apis/core/v1beta1/helper/helper.go [hunk 25]
```diff
@@ -1134,7 +1134,7 @@ func AnonymousAuthenticationEnabled(kubeAPIServerConfig *gardencorev1beta1.KubeA
 
 // CalculateSeedUsage returns a map representing the number of shoots per seed from the given list of shoots.
 // It takes both spec.seedName and status.seedName into account.
-func CalculateSeedUsage(shootList []gardencorev1beta1.Shoot) map[string]int {
+func CalculateSeedUsage(shootList []*gardencorev1beta1.Shoot) map[string]int {
```
Comment (timuthy) [inline pkg/apis/core/v1beta1/helper/helper.go:1137]: Do we still need this change?

**Review by timuthy — CHANGES_REQUESTED** [review]
Thanks for the adjustments. Seeing all the conversions, it presumably would have been better following the seconds approach, also because we could have dropped more helpers.
Any opinions @shafeeqes?
Nevertheless, there are probably some left-overs (see comments in-line).

**Comment by shafeeqes** [comment]
> Any opinions @shafeeqes?

No strong opinions, but I would slant towards the current state since it's been a lot of effort already.

File: plugin/pkg/shoot/validator/admission.go (hunk 23 above)
Comment (ary1992) [reply plugin/pkg/shoot/validator/admission.go:518]: `shootLister` returns the versioned shoot not the `core` shoot. It can be done but then we will have to convert the shootList to core shoot list.

File: plugin/pkg/managedseed/shoot/admission.go [hunk 26]
```diff
@@ -187,5 +187,11 @@ func (v *Shoot) getShoots(selector labels.Selector) ([]*core.Shoot, error) {
 	if err != nil {
 		return nil, apierrors.NewInternalError(err)
 	}
-	return shoots, nil
+
+	coreShoots, err2 := admissionutils.ConvertShootList(shoots)
+	if err2 != nil {
+		return nil, apierrors.NewInternalError(fmt.Errorf("could not convert v1beta1 shoot: %w", err2))
+	}
```
Comment (shafeeqes) [inline plugin/pkg/managedseed/shoot/admission.go]: Nit: no need of err2. You can name it err.

File: plugin/pkg/seed/validator/admission.go [hunk 27]
```diff
@@ -147,7 +147,12 @@ func (v *ValidateSeed) validateSeedDeletion(a admission.Attributes) error {
 		return apierrors.NewInternalError(err)
 	}
 
-	if admissionutils.IsSeedUsedByShoot(seedName, shoots) {
+	coreShoots, err2 := admissionutils.ConvertShootList(shoots)
+	if err2 != nil {
+		return apierrors.NewInternalError(fmt.Errorf("could not convert v1beta1 shoot: %w", err2))
+	}
```
Comment (shafeeqes) [inline plugin/pkg/seed/validator/admission.go]:
```suggestion
	if err != nil {
		return apierrors.NewInternalError(fmt.Errorf("could not convert v1beta1 shoot: %w", err))
	}
```

File: plugin/pkg/shoot/validator/admission.go [hunk 28]
```diff
@@ -239,33 +240,55 @@ func (v *ValidateShoot) Admit(ctx context.Context, a admission.Attributes, _ adm
...
 
 	// begin of validation code
 	validationContext := &validationContext{
-		cloudProfile:  cloudProfile,
+		cloudProfile:  coreCloudProfile,
 		project:       project,
-		seed:          seed,
-		secretBinding: secretBinding,
+		seed:          coreSeed,
+		secretBinding: coreSecretBinding,
```
Comment (shafeeqes) [inline plugin/pkg/shoot/validator/admission.go]: Naming is bit inconsistent, Should we just remove the `core` prefix? project is also internal type but doesn't have the prefix.

File: plugin/pkg/shoot/validator/admission.go [hunk 29]
```diff
@@ -497,13 +520,18 @@ func (c *validationContext) validateScheduling(ctx context.Context, a admission.
...
 	allShoots, err := shootLister.Shoots(metav1.NamespaceAll).List(labels.Everything())
 	if err != nil {
 		return 0, fmt.Errorf("could not list all shoots: %w", err)
 	}
 
-	seedUsage := helper.CalculateSeedUsage(allShoots)
+	coreShoots, err2 := admissionutils.ConvertShootList(allShoots)
+	if err2 != nil {
+		return 0, apierrors.NewInternalError(fmt.Errorf("could not convert v1beta1 shoot: %w", err2))
+	}
```
Comment (shafeeqes) [inline plugin/pkg/shoot/validator/admission.go]: same here

File: plugin/pkg/utils/miscellaneous.go [hunk 30]
```diff
@@ -46,13 +47,19 @@ func IsSeedUsedByShoot(seedName string, shoots []*core.Shoot) bool {
...
 	shoots, err := shootLister.List(labels.Everything())
 	if err != nil {
 		return nil, apierrors.NewInternalError(fmt.Errorf("failed to list shoots: %w", err))
 	}
-	for _, shoot := range shoots {
+
+	coreShoots, err2 := ConvertShootList(shoots)
+	if err2 != nil {
+		return nil, apierrors.NewInternalError(fmt.Errorf("could not convert v1beta1 shoot: %w", err2))
+	}
```
Comment (shafeeqes) [inline plugin/pkg/utils/miscellaneous.go]: same here

File: plugin/pkg/utils/miscellaneous.go [hunk 31]
```diff
@@ -77,17 +84,47 @@ func NewAttributesWithName(a admission.Attributes, name string) admission.Attrib
...
 		shoots, err := shootLister.List(labels.Everything())
 		if err != nil {
 			return err
 		}
 
-		if IsSeedUsedByShoot(seedName, shoots) {
+		coreShoots, err2 := ConvertShootList(shoots)
+		if err2 != nil {
+			return apierrors.NewInternalError(fmt.Errorf("could not convert v1beta1 shoot: %w", err2))
+		}
```
Comment (shafeeqes) [inline plugin/pkg/utils/miscellaneous.go]: same here

File: plugin/pkg/utils/miscellaneous.go [hunk 32]
```diff
@@ -77,17 +84,47 @@ func NewAttributesWithName(a admission.Attributes, name string) admission.Attrib
...
 			return apierrors.NewForbidden(core.Resource(kind), seedName, fmt.Errorf("cannot remove zones %v from %s %s as there are Shoots scheduled to this Seed", sets.List(removedZones), kind, seedName))
 		}
 	}
 
 	return nil
 }
+
+// ConvertList converts a list of elements from one type to another.
+// It takes a function f that performs the conversion for each element.
+func ConvertList[T, U any](list []T, f func(T) (U, error)) ([]U, error) {
```
Comment (shafeeqes) [inline plugin/pkg/utils/miscellaneous.go]: Nice!

File: plugin/pkg/shoot/validator/admission.go (hunk 28 above)
Comment (ary1992) [reply plugin/pkg/shoot/validator/admission.go]: There is no conversion needed for project hence it does not have the prefix. For the other types conversion was needed so I named it with `core` prefix

**Comment by timuthy** [comment]
/hold
There is currently an offline discussion about the current approach because of performance concerns.

File: plugin/pkg/managedseed/validator/admission.go [hunk 33]
```diff
@@ -517,18 +519,24 @@ func (v *ManagedSeed) getSeedDNSProviderForDefaultDomain(shoot *gardencore.Shoot
...
 
-func (v *ManagedSeed) getSecretBinding(namespace, name string) (*gardencore.SecretBinding, error) {
-	return v.secretBindingLister.SecretBindings(namespace).Get(name)
+func (v *ManagedSeed) getSecretBinding(namespace, name string) (*gardencorev1beta1.SecretBinding, error) {
+	secretBinding, err := v.secretBindingLister.SecretBindings(namespace).Get(name)
+	if err != nil {
+		return nil, err
+	}
+
+	return secretBinding, nil
```
Comment (shafeeqes) [inline plugin/pkg/managedseed/validator/admission.go]:
```suggestion
        return v.secretBindingLister.SecretBindings(namespace).Get(name)
```

File: plugin/pkg/shoot/quotavalidator/admission.go [hunk 34]
```diff
@@ -325,7 +326,12 @@ func (q *QuotaValidator) findShootsReferQuota(quota core.Quota, shoot core.Shoot
 				continue
 			}
 			if ptr.Deref(s.Spec.SecretBindingName, "") == binding.Name {
-				shootsReferQuota = append(shootsReferQuota, *s)
+				coreShoot := &core.Shoot{}
+				if err := gardencorev1beta1.Convert_v1beta1_Shoot_To_core_Shoot(s, coreShoot, nil); err != nil {
```
Comment (shafeeqes) [inline plugin/pkg/shoot/quotavalidator/admission.go:330]: Do we need this conversion? This is the shoot plugin, yes, the shoot is in internal version, but this shoot list is used to compare with the cloudprofile which is versioned, so isn't it better to usse v1beta1 shoots from the lister directly? WDYT?

File: plugin/pkg/shoot/quotavalidator/admission.go [hunk 35]
```diff
@@ -369,7 +375,12 @@ func (q *QuotaValidator) getShootResources(shoot core.Shoot) (corev1.ResourceLis
 		for _, e := range machineTypes {
 			element := e
 			if element.Name == worker.Machine.Type {
-				machineType = &element
+				coreElement := &core.MachineType{}
+				if err := gardencorev1beta1.Convert_v1beta1_MachineType_To_core_MachineType(&element, coreElement, nil); err != nil {
+					return nil, apierrors.NewInternalError(fmt.Errorf("could not convert machine type: %+v", err.Error()))
+				}
+
+				machineType = coreElement
```
Comment (shafeeqes) [inline plugin/pkg/shoot/quotavalidator/admission.go]: See comment above, we could probably avoid this conversion

File: plugin/pkg/shoot/quotavalidator/admission.go (hunk 34 above)
Comment (ary1992) [reply plugin/pkg/shoot/quotavalidator/admission.go:330]: Atleast at one place we will have to make the conversion.
Removing the conversion here, we will have to change the return type of this function to `gardencorev1beta1.shoot` and so on.

Comment (shafeeqes) [reply plugin/pkg/shoot/quotavalidator/admission.go:330]: okay, let's keep it then

File: pkg/apiserver/admission/initializer/types.go (hunk 10 above)
Comment (timuthy) [reply pkg/apiserver/admission/initializer/types.go]: I wanted to ask the same and then saw this comment. I don't know why the Kubernetes code wasn't adjusted but in our code it's called `WantsKubeClientset` as well (instead of `WantsExternalKubeClientSet`).

Hence, I would still vote dropping the "external" term here. We also don't have `WantsExternalSeedManagementClientSet` or similar.

File: plugin/pkg/shoot/dns/admission.go [hunk 36]
```diff
@@ -329,8 +329,8 @@ func assignDefaultDomainIfNeeded(shoot *core.Shoot, projectLister gardencorelist
 	return nil
 }
 
-func checkDefaultDomainFormat(a admission.Attributes, shoot *core.Shoot, projectLister gardencorelisters.ProjectLister, defaultDomains []string) error {
-	project, err := admissionutils.ProjectForNamespaceFromInternalLister(projectLister, shoot.Namespace)
+func checkDefaultDomainFormat(a admission.Attributes, shoot *core.Shoot, projectLister gardencorev1beta1listers.ProjectLister, defaultDomains []string) error {
+	project, err := admissionutils.ProjectForNamespaceFromExternalLister(projectLister, shoot.Namespace)
```
Comment (timuthy) [inline plugin/pkg/shoot/dns/admission.go]: Rename to `ProjectForNamespaceFromLister`.

**Review by timuthy — COMMENTED** [review]
Just a minor feedback, the rest looks good. Thank you 🙂

**Review by shafeeqes — COMMENTED** [review]
/lgtm

**Review by timuthy — COMMENTED** [review]
Nice, thanks for the effort and improvements 🙏 
/lgtm
/approve
